velox/common/file/File.cpp

```diff
@@ -304,26 +332,93 @@ void LocalWriteFile::append(std::unique_ptr<folly::IOBuf> data) {
   size_ += totalBytesWritten;
 }
 
+void LocalWriteFile::write(
+    const std::vector<iovec>& iovecs,
+    int64_t offset,
+    int64_t length) {
+  VELOX_CHECK(!closed_, "file is closed");
+  VELOX_CHECK_GE(offset, 0, "Offset cannot be negative.");
+  auto bytesWritten =
+      pwritev(fd_, iovecs.data(), static_cast<ssize_t>(iovecs.size()), offset);
+  VELOX_CHECK_EQ(
+      bytesWritten,
+      length,
+      "Failure in LocalWriteFile::write, {} vs {}",
+      bytesWritten,
+      length);
+  size_ = std::max<uint64_t>(size_, offset + bytesWritten);
+}
+
+void LocalWriteFile::truncate(int64_t newSize) {
+  VELOX_CHECK(!closed_, "file is closed");
+  VELOX_CHECK_GE(newSize, 0, "New size cannot be negative.");
+  auto rc = ftruncate(fd_, newSize);
+  VELOX_CHECK_EQ(
+      rc,
+      0,
+      "ftruncate failed in LocalWriteFile::truncate: {}.",
+      folly::errnoStr(errno));
+  size_ = newSize;
+}
+
 void LocalWriteFile::flush() {
   VELOX_CHECK(!closed_, "file is closed");
-  auto ret = fflush(file_);
+  auto rc = fsync(fd_);
   VELOX_CHECK_EQ(
-      ret,
+      rc,
       0,
-      "fflush failed in LocalWriteFile::flush: {}.",
+      "fsync failed in LocalWriteFile::flush: {}.",
       folly::errnoStr(errno));
 }
 
+void LocalWriteFile::setAttributes(int32_t attr) {
+  VELOX_CHECK(!closed_, "file is closed");
+#ifdef linux
+  const auto rc = ioctl(fd_, FS_IOC_SETFLAGS, &attr);
+  VELOX_CHECK_EQ(
+      0,
+      rc,
+      "ioctl(FS_IOC_SETFLAGS, attr {}) failed: {}, {}",
+      attr,
+      rc,
+      folly::errnoStr(errno));
+#endif // linux
+}
+
+int32_t LocalWriteFile::getAttributes() {
+  VELOX_CHECK(!closed_, "file is closed");
+  int attr{0};
+#ifdef linux
+  const auto rc = ioctl(fd_, FS_IOC_GETFLAGS, &attr);
```

Comment:
We can define a set of attributes which are specific to local file in LocalWriteFile as LocalWriteFile::kAttrXXX